File: app/src/main/java/com/example/agroapp/app/AppViewModelFactory.java
package com.example.agroapp.app;

import android.annotation.SuppressLint;
import android.app.Application;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.agroapp.repository.AgroRepository;
import com.example.agroapp.viewModel.DiscoverViewModel;
import com.example.agroapp.viewModel.HomeItemViewModel;
import com.example.agroapp.viewModel.HomeViewModel;
import com.example.agroapp.viewModel.NavigationViewModel;
import com.example.agroapp.viewModel.ProjectListViewModel;

/**
 * Created by goldze on 2019/3/26.
 */
public class AppViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    @SuppressLint("StaticFieldLeak")
    private static volatile AppViewModelFactory INSTANCE;
    private final Application mApplication;
    private final AgroRepository mRepository;


    public static AppViewModelFactory getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (AppViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new AppViewModelFactory(application, Injection.provideDemoRepository());
                }
            }
        }
        return INSTANCE;
    }

    @VisibleForTesting
    public static void destroyInstance() {
        INSTANCE = null;
    }

    private AppViewModelFactory(Application application, AgroRepository repository) {
        this.mApplication = application;
        this.mRepository = repository;

    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(HomeViewModel.class)) {
            return (T) new HomeViewModel(mApplication, mRepository);
        }else  if (modelClass.isAssignableFrom(ProjectListViewModel.class)) {
            return (T) new ProjectListViewModel(mApplication, mRepository);
        }else  if (modelClass.isAssignableFrom(NavigationViewModel.class)) {
            return (T) new NavigationViewModel(mApplication, mRepository);
        }else  if (modelClass.isAssignableFrom(DiscoverViewModel.class)) {
            return (T) new DiscoverViewModel(mApplication, mRepository);
        }


        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: kotlin/src/main/java/com/example/kotlin/app/AppViewModelFactory.kt
package com.example.kotlin.app

import android.annotation.SuppressLint
import android.app.Application
import androidx.annotation.VisibleForTesting
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.agroapp.repository.AgroRepository
import com.example.agroapp.viewModel.DiscoverViewModel
import com.example.agroapp.viewModel.HomeViewModel
import com.example.agroapp.viewModel.NavigationViewModel
import com.example.agroapp.viewModel.ProjectListViewModel

/**
 * Created by goldze on 2019/3/26.
 */
class AppViewModelFactory private constructor(private val mApplication: Application, repository: AgroRepository) : ViewModelProvider.NewInstanceFactory() {
    private val mRepository: AgroRepository
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(HomeViewModel::class.java)) {
            return HomeViewModel(mApplication, mRepository) as T
        } else if (modelClass.isAssignableFrom(ProjectListViewModel::class.java)) {
            return ProjectListViewModel(mApplication, mRepository) as T
        } else if (modelClass.isAssignableFrom(NavigationViewModel::class.java)) {
            return NavigationViewModel(mApplication, mRepository) as T
        } else if (modelClass.isAssignableFrom(DiscoverViewModel::class.java)) {
            return DiscoverViewModel(mApplication, mRepository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class: " + modelClass.name)
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        @Volatile
        private var INSTANCE: AppViewModelFactory? = null
        fun getInstance(application: Application): AppViewModelFactory? {
            if (INSTANCE == null) {
                synchronized(AppViewModelFactory::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = AppViewModelFactory(application, Injection.provideDemoRepository())
                    }
                }
            }
            return INSTANCE
        }

        @VisibleForTesting
        fun destroyInstance() {
            INSTANCE = null
        }
    }

    init {
        mRepository = repository
    }
}